File: ILT.IHR.UI/Service/LeaveBalanceService.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Net.Http;
using System.Threading.Tasks;
using Blazored.SessionStorage;
using ILT.IHR.DTO;
using ILT.IHR.UI.ExtensionMethods;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Authorization;
using Newtonsoft.Json;

namespace ILT.IHR.UI.Service
{
    public class LeaveBalanceService: ILeaveBalanceService
    {
        private readonly HttpClient httpClient;
        private ISessionStorageService _sessionStorageService;
        [Inject]
        public AuthenticationStateProvider _authenticationStateProvider { get; set; }
        public LeaveBalanceService(HttpClient httpClient, ISessionStorageService sessionStorageService, AuthenticationStateProvider authenticationStateProvider)
        {
            this.httpClient = httpClient;
            _sessionStorageService = sessionStorageService;
            _authenticationStateProvider = authenticationStateProvider;
        }
        public async Task<Response<IEnumerable<LeaveBalance>>> GetLeaveBalance(int? EmployeID)
        {
            var token = await _sessionStorageService.GetItemAsync<string>("token");
            httpClient.SetAuthorizationHeader(token);
            try
            {
                if (EmployeID == 0)
                {
                    return await httpClient.GetJsonAsync<Response<IEnumerable<LeaveBalance>>>("LeaveBalance");
                }
                return await httpClient.GetJsonAsync<Response<IEnumerable<LeaveBalance>>>($"LeaveBalance/?EmployeeID={EmployeID}");
            }
            catch (Exception e)
            {
                ((Data.CustomAuthenticationStateProvider)_authenticationStateProvider).SessionExpired(e.ToString());
                return new Response<IEnumerable<LeaveBalance>>();
            }
        }
        public async Task<Response<LeaveBalance>> GetLeaveBalanceById(int ID)
        {
            var token = await _sessionStorageService.GetItemAsync<string>("token");
            httpClient.SetAuthorizationHeader(token);
            try
            {
                return await httpClient.GetJsonAsync<Response<LeaveBalance>>($"LeaveBalance/{ID}");
            }
            catch (Exception e)
            {
                ((Data.CustomAuthenticationStateProvider)_authenticationStateProvider).SessionExpired(e.ToString());
                return new Response<LeaveBalance>();
            }
        }
        public async Task<Response<LeaveBalance>> updateLeaveBalance(int Id, LeaveBalance updateObject)
        {
            var token = await _sessionStorageService.GetItemAsync<string>("token");
            httpClient.SetAuthorizationHeader(token);
            try
            {
                return await httpClient.PutJsonAsync<Response<LeaveBalance>>($"LeaveBalance/{Id}", updateObject);
            }
            catch (Exception e)
            {
                ((Data.CustomAuthenticationStateProvider)_authenticationStateProvider).SessionExpired(e.ToString());
                return new Response<LeaveBalance>();
            }
        }
        public async Task<DataTable> GetReportLeaveInfo(DTO.Report reportReq, string LeaveSummaryStatus)
        {
            var token = await _sessionStorageService.GetItemAsync<string>("token");
            httpClient.SetAuthorizationHeader(token);
            try
            {
                var resp = await httpClient.PostJsonAsync<Response<List<LeaveBalance>>>("LeaveBalance/GetLeavesCount", reportReq);
                if (!String.IsNullOrEmpty(LeaveSummaryStatus) && resp.Data != null)
                {
                    if (LeaveSummaryStatus == "All")
                    {
                        string json = JsonConvert.SerializeObject(resp.Data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                    else if (LeaveSummaryStatus != "Active")
                    {
                        var data = resp.Data.Where(x => x.TermDate != null && x.TermDate <= DateTime.Now).ToList();
                        string json = JsonConvert.SerializeObject(data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                    else
                    {
                        var data = resp.Data.Where(x => x.TermDate == null || x.TermDate > DateTime.Now).ToList();
                        string json = JsonConvert.SerializeObject(data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                }
                else
                {
                    string json = JsonConvert.SerializeObject(resp.Data);
                    DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                    return dt;
                }
            }
            catch (Exception e)
            {
                ((Data.CustomAuthenticationStateProvider)_authenticationStateProvider).SessionExpired(e.ToString());
                return new DataTable();
            }
        }
        public async Task<DataTable> GetReportLeaveDetailInfo(DTO.Report reportReq, string LeaveDetailStatus)
        {
            var token = await _sessionStorageService.GetItemAsync<string>("token");
            httpClient.SetAuthorizationHeader(token);
            try
            {
                var resp = await httpClient.PostJsonAsync<Response<List<LeaveBalance>>>("LeaveBalance/GetLeaveDetail", reportReq);
                if (!String.IsNullOrEmpty(LeaveDetailStatus) && resp.Data != null)
                {
                    if (LeaveDetailStatus == "All")
                    {
                        string json = JsonConvert.SerializeObject(resp.Data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                    else if (LeaveDetailStatus != "Active")
                    {
                        var data = resp.Data.Where(x => x.TermDate != null && x.TermDate <= DateTime.Now).ToList();
                        string json = JsonConvert.SerializeObject(data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                    else
                    {
                        var data = resp.Data.Where(x => x.TermDate == null || x.TermDate > DateTime.Now).ToList();
                        string json = JsonConvert.SerializeObject(data);
                        DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                        return dt;
                    }
                } else
                {
                    string json = JsonConvert.SerializeObject(resp.Data);
                    DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof(DataTable)));
                    return dt;
                }
            }
            catch (Exception e)
            {
                ((Data.CustomAuthenticationStateProvider)_authenticationStateProvider).SessionExpired(e.ToString());
                return new DataTable();
            }
        }
    }
}

File: ILT.IHR.APP/ClientApp/src/app/pages/manage-leave/LeaveBalanceService.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ILeaveBalance } from '../../core/interfaces/LeaveBalance';

@Injectable({
    providedIn: 'root'
})
export class LeaveBalanceService {
    constructor(private http: HttpClient) { }
    getGetLeaveBalanceApiMethod = "api/LeaveBalance";
    getGetLeaveBalanceByIDApiMethod = "api/LeaveBalance";
    updateLeaveBalanceApiMethod = "api/LeaveBalance";
    GetReportLeaveDetailInfoApiMethod = "api/LeaveBalance/GetLeaveDetail";

    GetLeaveBalance(employeeId: number) {
        if (employeeId === 0) {
            return this.http.get<any>(`${this.getGetLeaveBalanceApiMethod}`);
        } else {
            return this.http.get<any>(`${this.getGetLeaveBalanceApiMethod}?EmployeeID=${employeeId}`);
        }
    }

    GetLeaveBalanceById(ID: number) {
        return this.http.get<any>(`${this.getGetLeaveBalanceByIDApiMethod}/${ID}`);
    }
    UpdateLeaveBalance(ID: number, leaveBalance: ILeaveBalance): Observable<any[]> {
        return this.http.put<any>(`${this.updateLeaveBalanceApiMethod}/${ID}`, leaveBalance);
    }
    GetReportLeaveDetailInfo(reportReq: any, LeaveDetailStatus: any) {
        console.log('Heree', reportReq)
      return this.http.post<any>(`${this.GetReportLeaveDetailInfoApiMethod}`, reportReq);
        //console.log(resp)
        //if ((LeaveDetailStatus !== null && LeaveDetailStatus !== '') && resp['data'] !== null) {
        //    if (LeaveDetailStatus == "All") {
        //        var json = resp['data'];
        //        //  DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof (DataTable)));
        //        return json;
        //    } else {
        //         var json = resp['data'];
        //        //  DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof (DataTable)));
        //        return json;
        //    }
        //    //else if (LeaveDetailStatus != "Active") {
        //    //    var data = resp.Data.Where(x => x.TermDate != null && x.TermDate <= DateTime.Now).ToList();
        //    //    string json = JsonConvert.SerializeObject(data);
        //    //    DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof (DataTable)));
        //    //    return dt;
        //    //}
        //    //else {
        //    //    var data = resp.Data.Where(x => x.TermDate == null || x.TermDate > DateTime.Now).ToList();
        //    //    string json = JsonConvert.SerializeObject(data);
        //    //    DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof (DataTable)));
        //    //    return dt;
        //    //}
        //} else {
        //    var json = resp['data'];
        //    console.log(json)
        //    //string json = JsonConvert.SerializeObject(resp.Data);
        //    //DataTable dt = (DataTable)JsonConvert.DeserializeObject(json, (typeof (DataTable)));
        //    //return dt;
        //}

    }
}
